File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**9)

n, m = map(int,input().split())
graph = [[] for _ in range(n+1)]
visited = [False] * (n+1)
connect = 0

def dfs(x):
    visited[x] = True
    for i in graph[x]:
        if visited[i] == False:
            dfs(i)

for i in range(m):
    v1, v2 = map(int, input().split())
    graph[v1].append(v2)
    graph[v2].append(v1)

for i in range(1, n+1):
    if visited[i] == False:
        dfs(i)
        connect += 1

print(connect)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const [n, m] = input[0].split(' ').map(Number)
const graph = Array.from({ length: n + 1 }, () => [])
for (let i = 1; i < m + 1; i++) {
  const [a, b] = input[i].split(' ').map(Number)
  graph[a].push(b)
  graph[b].push(a)
}

const visited = Array.from({ length: n + 1 }, () => false)
let cnt = 0
for (let i = 1; i < n + 1; i++) {
  if (visited[i]) continue

  cnt++
  bfs(i)
}

console.log(cnt)

function bfs(start) {
  let idx = 0
  const queue = []

  visited[start] = true
  queue.push(start)

  while(idx < queue.length) {
    const node = queue[idx++]

    for (const nextNode of graph[node]) {
      if (visited[nextNode]) continue

      visited[nextNode] = true
      queue.push(nextNode)
    }
  }
}